File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length();
        HashSet<Character> set = new HashSet<>();
        int l =0,r=0;
        char ch;
        int ans =0;
        while(r < n){
            ch = s.charAt(r);
            while(set.contains(ch)){
                set.remove(s.charAt(l));
                l++;
            }
            set.add(ch);
            ans = Math.max(ans,r-l+1);
            r++;
        }
        return ans;
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = s.size();
        map<int, int> map;
        int ans = 0;
        int l =0,r=0;
        while(r < n){
            char ch = s[r];
            if(map.find(ch) != map.end()){
                l = max(l, map[ch]+1);
            }

            map[ch] = r;
            ans = max(ans , r-l+1);
            r++;
        }
        return ans;
    }
};